File: structural/typescript/flyweight_structured/game_soldiers.ts
import { SoldierType } from './interfaces';

export class ConcreteSoldierType implements SoldierType {
  constructor(
    private uniform: string,
    private speed: number,
    private rank: string
  ) {}
  
  render(x: number, y: number, weapon: string): void {
    console.log(`Rendering ${this.rank} soldier at (${x}, ${y}) with ${this.uniform} uniform, speed ${this.speed}, holding ${weapon}`);
  }
  
  getUniform(): string {
    return this.uniform;
  }
  
  getSpeed(): number {
    return this.speed;
  }
}

export class SoldierTypeFactory {
  private soldierTypes: Map<string, SoldierType> = new Map();
  
  private getKey(uniform: string, speed: number, rank: string): string {
    return `${uniform}|${speed}|${rank}`;
  }
  
  getSoldierType(uniform: string, speed: number, rank: string): SoldierType {
    const key = this.getKey(uniform, speed, rank);
    
    if (!this.soldierTypes.has(key)) {
      console.log(`Creating new soldier type: ${rank}`);
      this.soldierTypes.set(key, new ConcreteSoldierType(uniform, speed, rank));
    } else {
      console.log(`Reusing existing soldier type: ${rank}`);
    }
    
    return this.soldierTypes.get(key)!;
  }
  
  getTotalSoldierTypes(): number {
    return this.soldierTypes.size;
  }
}

export class Soldier {
  constructor(
    private x: number,
    private y: number,
    private weapon: string,
    private soldierType: SoldierType
  ) {}
  
  render(): void {
    this.soldierType.render(this.x, this.y, this.weapon);
  }
  
  move(newX: number, newY: number): void {
    this.x = newX;
    this.y = newY;
  }
}

File: structural/dart/flyweight_structured/game_soldiers.dart
import 'interfaces.dart';

class ConcreteSoldierType extends SoldierType {
  @override
  final String uniform;
  @override
  final int speed;
  final String rank;
  
  ConcreteSoldierType(this.uniform, this.speed, this.rank);
  
  @override
  void render(int x, int y, String weapon) {
    print('Rendering $rank soldier at ($x, $y) with $uniform uniform, speed $speed, holding $weapon');
  }
}

class SoldierTypeFactory {
  final Map<String, SoldierType> _soldierTypes = {};
  
  String _getKey(String uniform, int speed, String rank) {
    return '$uniform|$speed|$rank';
  }
  
  SoldierType getSoldierType(String uniform, int speed, String rank) {
    var key = _getKey(uniform, speed, rank);
    
    if (!_soldierTypes.containsKey(key)) {
      print('Creating new soldier type: $rank');
      _soldierTypes[key] = ConcreteSoldierType(uniform, speed, rank);
    } else {
      print('Reusing existing soldier type: $rank');
    }
    
    return _soldierTypes[key]!;
  }
  
  int getTotalSoldierTypes() {
    return _soldierTypes.length;
  }
}

class Soldier {
  int x, y;
  final String weapon;
  final SoldierType soldierType;
  
  Soldier(this.x, this.y, this.weapon, this.soldierType);
  
  void render() {
    soldierType.render(x, y, weapon);
  }
  
  void move(int newX, int newY) {
    x = newX;
    y = newY;
  }
}
